grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy

```diff
@@ -0,0 +1,158 @@
+     * continuable.
+     */
+    @Override
+    RepeatStatus execute(StepContribution contribution, ChunkContext chunkContext) throws Exception {
+        final Session session = JcrUtil.getSession(slingRepository, "admin")
+        //If we don't have any exclude paths, we don't have to worry about only clearing particular subtrees
+        if(!excludePaths || excludePaths.empty) {
+            final Node node = session.getNode(jobPath)
+            node.remove()
+        }
+        else {
+            deletePartialTree(session)
+        }
+        session.save()
+        return RepeatStatus.FINISHED
+    }
+
+
+    private void deletePartialTree(final Session session) {
+        final relativeExcludePaths = excludePaths.collect { it - jobPath }
+        final relativeRoot = session.getNode(jobPath)
+        deletePartialTree(relativeRoot, relativeExcludePaths)
+    }
+
+
+    private void deletePartialTree(Node rootNode, Collection<String> relativeExcludePaths) {
+        //Base case.  If a root node does not have any exclude paths under it, it must have been excluded
+        if(!relativeExcludePaths) {
+            return
+        }
+        //Compute the current tree level excluded nodes, and compute the remaining paths to traverse
+        Collection<NodeAndExcludePaths> nodeAndExcludePaths = relativeExcludePaths.inject([] as Collection<NodeAndExcludePaths>) { def acc, def thisPath ->
+            final thisNodeAndExcludePath = NodeAndExcludePaths.fromPath(thisPath)
+            //If we have already created a nodeAndExcludePath for this node name, just add to it's exclude paths
+            final matchingNodeName = acc.find { it.nodeName == thisNodeAndExcludePath.nodeName }
+            if(matchingNodeName) {
+                matchingNodeName.excludePaths.addAll(thisNodeAndExcludePath.excludePaths)
+            }
+            else {
+                acc.add(thisNodeAndExcludePath)
+            }
+            return acc
+        } as Collection
+        //Delete nodes allowed under this tree
+        final rootNodeChildren = rootNode.getNodes()
+        while(rootNodeChildren.hasNext()) {
+            final currentNode = rootNodeChildren.nextNode()
+            //If this node is in our exclusion list, don't delete it
+            if(!(nodeAndExcludePaths.find { currentNode.name == it.nodeName })) {
+                currentNode.remove()
+            }
+        }
+        //Recurse on each
+        nodeAndExcludePaths.each {
+            deletePartialTree(rootNode.getNode(it.nodeName), it.excludePaths)
+        }
+    }
+
+    /**
+     * If we have a path for a node such as 'foo/bar/doo'
+     * node is 'foo', and the excludePaths is '/bar/doo'
+     */
+    static class NodeAndExcludePaths {
+
+        String nodeName
+        Collection<String> excludePaths
+
+        static NodeAndExcludePaths fromPath(final String path) {
+            //Remove leading '/'
+            final String thisPath = path.replaceFirst('/', '')
```

Comment:
Actually, given the example in the javadoc, this would mess things up.

"foo/bar/doo" -> {"foobar", "/doo"}